lib/datasource.js

```diff
@@ -365,36 +365,60 @@ DataSource.prototype.setup = function(name, settings) {
     }
   }
 
-  dataSource.connect = function(cb) {
-    var dataSource = this;
-    if (dataSource.connected || dataSource.connecting) {
-      process.nextTick(function() {
-        cb && cb();
-      });
+  dataSource.connect = function(callback) {
+    var self = this;
+    if (this.connected) {
+      // It's already connected or being connected
+      // Do NOT run loadFromFile again as it will override the cache
+      if (callback) process.nextTick(callback);
       return;
     }
-    dataSource.connecting = true;
-    if (dataSource.connector.connect) {
-      dataSource.connector.connect(function(err, result) {
-        if (!err) {
-          dataSource.connected = true;
-          dataSource.connecting = false;
-          dataSource.emit('connected');
-        } else {
-          dataSource.connected = false;
-          dataSource.connecting = false;
-          dataSource.emit('error', err);
-        }
-        cb && cb(err, result);
-      });
-    } else {
+    if (typeof dataSource.connector.connect !== 'function') {
+      // Connector doesn't have the connect function
+      // Assume no connect is needed
       process.nextTick(function() {
-        dataSource.connected = true;
-        dataSource.connecting = false;
-        dataSource.emit('connected');
-        cb && cb();
+        self.connected = true;
+        self.connecting = false;
+        self.emit('connected');
+        callback && callback();
       });
+      return;
     }
+
+    if (callback) {
+      // Queue the callback
+      this.pendingConnectCallbacks = this.pendingConnectCallbacks || [];
+      this.pendingConnectCallbacks.push(callback);
+    }
+    // The connect is already in progress
+    if (this.connecting) return;
+    // Set connecting flag to be true
+    this.connecting = true;
+    this.connector.connect(function(err, result) {
+      self.connecting = false;
+      if (!err) self.connected = true;
+      var cbs = self.pendingConnectCallbacks;
+      self.pendingConnectCallbacks = [];
+      if (!err) {
+        self.emit('connected');
+      } else {
+        self.emit('error', err);
+      }
+      // Invoke all pending callbacks
+      async.each(cbs, function(cb, done) {
+        process.nextTick(function() {
+          try {
+            cb(err);
+          } catch (e) {
+            // Ignore error to make sure all callbacks are invoked
```

Comment:
This will silently swallow errors that may be important for troubleshooting. I am proposing the following implementation:

```js
async.each(cbs, (cb, next) => {
  process.nextTick(cb, err);
  next();
});
```

That way any errors thrown by the callback trigger "unhandled error" handling, which typically crashes the application (as expected for an unhandled error).